Use libmkl_rt, or statically link against MKL

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #31162 Don't use RTLD_GLOBAL to load _C.
* #31165 Pick parallel MKL implementation over sequential implementation.
* **#31159 Use libmkl_rt, or statically link against MKL**
* #31161 Actually get rid of Windows hacks in cpp_extensions.
* #31160 Stop using ctypes to interface with CUDA libraries.
* #31176 Move AutogradMeta and DeviceGuardImplInterface virtual methods out-of-line.
* #31157 Add missing TORCH_CUDA_API annotation to throw_nccl_error
* #31155 Remove dead CAFFE2_LIBS variable
* #31152 Remove LibIRC logic from cmake.

Part of removing RTLD_GLOBAL for _C.

MKL doesn't like being in a local scope, unless you use the libmkl_rt.so
variant, so I adapted our CMake configuration to either accept that or
only allow static linking.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Use libmkl_rt, or statically link against MKL

Part of removing RTLD_GLOBAL for _C.

MKL doesn't like being in a local scope, unless you use the libmkl_rt.so
variant, so I adapted our CMake configuration to either accept that or
only allow static linking.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

kostmo commented:
## CircleCI build failures summary
As of commit 4351fec:
* **1/1** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=4351fec2ca1889310a0d460ad3fdb6af75f24098).

### 1 failure *not* recognized by patterns:
| Job | Step |
| --- | --- |
| ![CircleCI](https://avatars0.githubusercontent.com/ml/7?s=12) <sup>pytorch_macos_10_13_py3_test</sup> | <sup>Test</sup> |


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

ezyang commented:
Hooo boy, this is the change that's causing the numeric errors.

ezyang committed:
Update on "Use libmkl_rt, or statically link against MKL"


Part of removing RTLD_GLOBAL for _C.

MKL doesn't like being in a local scope, unless you use the libmkl_rt.so
variant, so I adapted our CMake configuration to either accept that or
only allow static linking.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]
